addons/udes_stock/tests/test_push_from_drop.py

```diff
@@ -0,0 +1,253 @@
+        Move = self.env['stock.move']
+        res = Move._create_moves_for_push(self.push_damaged_putaway,
+                                          self.move_lines)
+        self.assertEqual(len(res), 2)
+        new_apple_move = res.filtered(lambda m: m.product_id.id == self.apple.id)
+        new_banana_move = res.filtered(lambda m: m.product_id.id == self.banana.id)
+
+        self.assertEqual(new_apple_move.product_uom_qty, 10)
+        self.assertEqual(new_banana_move.product_uom_qty, 99)
+
+    def test02_correct_move_locations(self):
+        """Test that created moves have the correct src and dest locations"""
+        Move = self.env['stock.move']
+        res = Move._create_moves_for_push(self.push_damaged_putaway,
+                                          self.move_lines)
+
+        self.assertEqual(len(res), 2)
+        for move in res:
+            self.assertEqual(move.location_id, self.received_damaged_location)
+            self.assertEqual(move.location_dest_id, self.location_qc_zone)
+
+    def test03_correct_move_orig_info(self):
+        """Test that created moves set move_orig_ids correctly"""
+        Move = self.env['stock.move']
+
+        moves = self.goods_in.mapped('move_lines')
+        apple_move = moves.filtered(lambda m: m.product_id.id == self.apple.id)
+        banana_move = moves.filtered(lambda m: m.product_id.id == self.banana.id)
+        self.assertFalse(apple_move.move_dest_ids)
+        self.assertFalse(banana_move.move_dest_ids)
+
+        res = Move._create_moves_for_push(self.push_damaged_putaway,
+                                          self.move_lines)
+
+        self.assertEqual(len(res), 2)
+        new_apple_move = res.filtered(lambda m: m.product_id.id == self.apple.id)
+        new_banana_move = res.filtered(lambda m: m.product_id.id == self.banana.id)
+        self.assertEqual(new_apple_move.move_orig_ids, apple_move)
+        self.assertEqual(new_banana_move.move_orig_ids, banana_move)
+
+
+class TestPushFromDrop(PushFromDropBase):
+    """Tests for the full push from drop functionality"""
+    def setUp(self):
+        """Create a picking to work with for tests"""
+        super(TestPushFromDrop, self).setUp()
+        products_info = [{'product': self.apple, 'uom_qty': 10},
+                         {'product': self.banana, 'uom_qty': 99}]
+        self.goods_in = self.create_picking(self.picking_type_goods_in,
+                                            products_info=products_info)
+        self.goods_in.action_confirm()
+        self.goods_in.action_assign()
+        self.move_lines = self.goods_in.move_line_ids
+        for ml in self.move_lines:
+            ml.write({'qty_done': ml.product_uom_qty})
+
+    def test01_picking_has_correct_src_and_dest(self):
+        """Test that pickings created by push from drop have the src and dest
+        location of the rule that created them."""
+        for ml in self.move_lines:
+            ml.write({'qty_done': ml.product_uom_qty,
```

Comment:
Why are we re-writing the qty_done here - we have done it already in the setUp method.
We could write the dest location here as well, then do caleb's suggestion of next_picking_ids in the setup :)

Reply:
We may not be able to do it in the setup method because of the unlinking in test02_no_push_rule, 
could add a method instead 
_complete_picking or something similar, and can call action_done there and return the next pickings (possibly)